# [KDT] ch1 데이타분석용 파이썬 이해

In [10]:
import os
os.getcwd()

'/Users/kelvin/Desktop/workspace/AI_Basic'

In [11]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [42]:
pd.__version__
np.__version__


'1.24.2'

# 1. Numpy

In [4]:
#import numpy 
import numpy as np

In [22]:
# nd array 1차원
np.array([1,2,3,4,5]).shape

(5,)

In [26]:
# nd array 2차원
np.array([[ 1,  3,  5,  7,  9,11, 13, 15, 17, 19, 21, 23, 25, 27, 29]]).shape

(1, 15)

In [43]:
# nd array 3차원
np.array([[[ 1,  3,  5,  7,  9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29]]]).shape

(1, 1, 15)

### 1_1. ndarray의 데이터 타입

In [30]:
# int type
list1 = [1,2,3]
list1
print(type(list))
array1 = np.array(list1)
array1
print(type(array1))
print(array1, array1.dtype)

<class 'type'>
<class 'numpy.ndarray'>
[1 2 3] int64


In [35]:
# int type + object
list2 = [1,2,3,4.0]
list2
array2 = np.array(list2)
array2
print(array2,array2.dtype)
# int type + float


[1. 2. 3. 4.] float64


In [41]:
# type 변환, astype('float'), astype('int')
array_int = np.array([1,2,3])
array_float = np.array([1.0, 2.0, 3.0])
array_float.dtype

array_ch = array_int.astype(float)
array_ch.dtype

dtype('float64')

### 1_2. ndarray를 편리하게 생성하기 - arange, zeros, ones

In [45]:
# np.arange
sequence_array = np.arange(10)
sequence_array.dtype
sequence_array.shape

(10,)

In [50]:
#(3, 2) shape을 가지는 모든 원소가 0, dtype은 int32 인 ndarray 생성. 
zero_array = np.zeros((3,2),dtype=int)
zero_array.dtype
zero_array.shape
zero_array
#(3, 2) shape을 가지는 모든 원소가 1인 ndarray 생성.
ones_array = np.ones((3,2))
ones_array.dtype
ones_array.shape

(3, 2)

### 1_3. ndarray reshape

In [59]:
# reshape
array1 = np.arange(10)
array1
# (2, 5) shape으로 변환
array2 = array1.reshape(2,5)
#(5, 2) shape으로 변환
array3 = array1.reshape(5,2)

In [54]:
# reshape (X)
array4 = array3.reshape(4,3)

NameError: name 'array3' is not defined

In [60]:
# reshape(-1)
array5 = array3.reshape(2,-1)
array5

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [61]:
# reshape (X)
array5 = array3.reshape(-1,4)

ValueError: cannot reshape array of size 10 into shape (4)

In [67]:
# reshape (ndarray 차원변환) 
array1 = np.arange(8)
print('array1:\n',array1.tolist())
print('array1:\n',array1)
# 1차원 ndarray를 3차원 ndarray로 변환 
array3d = array1.reshape((2,2,2))
print('array3d:\n',array3d.tolist())
print('array3d:\n',array3d)
# 3차원 ndarray를 2차원 ndarray로 변환하되 칼럼갯수는 1 
array5 = array3d.reshape(-1,1)
print('array5:\n',array5.tolist())
print('array5:\n',array5)
# 1차원 ndarray를 2차원 ndarray로 변환화되 칼럼 갯수는 1 

# 3차원 ndarray를 1차원으로 변환 

array1:
 [0, 1, 2, 3, 4, 5, 6, 7]
array1:
 [0 1 2 3 4 5 6 7]
array3d:
 [[[0, 1], [2, 3]], [[4, 5], [6, 7]]]
array3d:
 [[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]
array5:
 [[0], [1], [2], [3], [4], [5], [6], [7]]
array5:
 [[0]
 [1]
 [2]
 [3]
 [4]
 [5]
 [6]
 [7]]


### 1_4. ndarray indexing

* 단일 값 추출 (1d array)

In [70]:
# 1에서 부터 12 까지의 1차원 ndarray 생성 
arr = np.arange(12)
print('arr=',arr)

arr= [ 0  1  2  3  4  5  6  7  8  9 10 11]


In [72]:
# array[-1] 맨 뒤값
print(arr[-1])

11


단일 값 추출 (2d array)

In [74]:
# 실습 array 생성
arr = np.arange(12)
arr2 = arr.reshape(2,-1)

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11]])

In [83]:
# ndarray indexing
arr = np.arange(12)
arr2 = arr.reshape(2,-1)
print(arr2[1,5])

11


* 슬라이싱 - 연속된 인덱스상의 값을 추출하는 방식

In [ ]:
# 실습 array 생성
arr = np.arange(12)
arr2 = arr.reshape(2,-1)

In [85]:
# 슬라이싱 1d
arr[0:5]

array([0, 1, 2, 3, 4])

In [87]:
# 슬라이싱 2d
arr2[1,1:5]

array([ 7,  8,  9, 10])

* 팬시 인덱싱 - 리스트나 ndarray로 인덱스 집합을 지정하는 방식

In [89]:
# 실습 array 생성
arr_12 = np.arange(12)
print('arr_12\n', arr_12)
arr_26 = arr_12.reshape(2,6)
print('arr_26\n', arr_26)
arr_34 = arr_12.reshape(3,-1)
print('arr_34\n', arr_34)

arr_12
 [ 0  1  2  3  4  5  6  7  8  9 10 11]
arr_26
 [[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]]
arr_34
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


In [91]:
# 팬시 인덱싱 지정할 수 있음
arr_34[[0,1],0:2]

array([[0, 1],
       [4, 5]])

* 불린 인덱싱- True, False 조건으로 인덱스 집합을 지정하는 방식

In [ ]:
# 실습 array 생성
arr_12 = np.arange(12)
print('arr_12\n', arr_12)

In [93]:
# 불린 인덱싱
arr_12 > 4

array([False, False, False, False, False,  True,  True,  True,  True,
        True,  True,  True])

In [95]:
# 불린 인덱싱, boolean_indexes => boolean_indexes는 필터링에 활용
arr_12[arr_12 > 4]

array([ 5,  6,  7,  8,  9, 10, 11])

In [96]:
# arr_12를 2로 나눈 나머지가 짝수인가
np.mod(arr_12,2) == 0

array([ True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False])

In [99]:
# arr_12에서 홀수만 추출(필터링)
boolean_indexer = [ True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False]
arr_12[boolean_indexer]

array([ 0,  2,  4,  6,  8, 10])

연습문제 풀이

# 2. Pandas

### 2_1. DataFrames 생성

* pd.DataFrame() 

In [102]:
# list로 생성
import pandas as pd
list = [[1,2,3,4],[5,6,7,8]]
list
pd.DataFrame(list)


0  1  2  3
0  1  2  3  4
1  5  6  7  8

In [104]:
# list로 생성, columnlumn 추가
pd.DataFrame(list, columns=['A','B','C','D'])

A  B  C  D
0  1  2  3  4
1  5  6  7  8

In [105]:
# list로 생성, column 추가, indexex 추가
pd.DataFrame(list, columns=['A','B','C','D'], index = ['one','two'])

A  B  C  D
one  1  2  3  4
two  5  6  7  8

In [112]:
# array로 생성
arr = np.arange(12)
arr2 = arr.reshape(3,-1)
arr2
pd.DataFrame(arr2, columns=['A','B','C','D'], index = ['one','two','three'])

A  B   C   D
one    0  1   2   3
two    4  5   6   7
three  8  9  10  11

In [111]:
# array로 생성, column추가, index 추가
pd.DataFrame(arr2, columns=['A','B','C','D'], index = ['one','two','three'])

A  B   C   D
one    0  1   2   3
two    4  5   6   7
three  8  9  10  11

In [113]:
# 딕셔너리로 생성
dict1 = {'A':[1,2,3],'B':[4,5,6],'C':[7,8,9]}

In [114]:
# 딕셔너리로 생성, index 추가
pd.DataFrame(dict1,index =['one','two','three'])

A  B  C
one    1  4  7
two    2  5  8
three  3  6  9

* DataFrame을 넘파이 ndarray, list, 딕셔너리로 변환하기

In [115]:
# DataFrame 생성
df = pd.DataFrame(dict1,index =['one','two','three'])

In [117]:
# DataFrame을 딕셔너리로 변환
df.to_dict

<bound method DataFrame.to_dict of        A  B  C
one    1  4  7
two    2  5  8
three  3  6  9>

In [123]:
# DataFrame을 ndarray로 변환
df.values

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [126]:
# DataFrame을 list로 변환
df.values.tolist()

[[1, 4, 7], [2, 5, 8], [3, 6, 9]]

## Quiz 1

1) 아래 DataFrame을 딕셔너리로 생성하시오

In [127]:
# pd.DataFrame({ })
dict2 = {'name':['kdh100','kdh101','kdh101','kdh103'],'jumin7':['990623-2','900303-2','681205-2','931226-2'],
         'sex':['F','F','F','F']}
pd.DataFrame(dict2)

name    jumin7 sex
0  kdh100  990623-2   F
1  kdh101  900303-2   F
2  kdh101  681205-2   F
3  kdh103  931226-2   F

### 2_2. Data Frame 읽기
* read_csv()** csv, txt, Excel Data 가능


* 데이터 읽어오기 pd.read_csv('file_name.csv')
- [데이터 읽어오기 참고(csv,txt,xlsx,UTF-8 인코딩)]
- df = pd.read_csv('file_name.csv')
- txt: 텍스트 파일. 특정 구분자로 구분되어있다면 sep 인자에 명시(탭(tab)은 sep='\t', '|' 는 sep='|'
- df = pd.read_csv('file_name.csv', encoding='cp949') - 한글파일 인코딩 에러시
- df = pd.read_csv('file_name.csv', sep='\t') - 구분자: Tab일때
                                        
- xlsx: 엑셀 파일. 특정 시트(sheet)의 데이터만 읽어올 경우 sheet_name 인자에 시트명 또는 순번을 입력해준다.
- df = pd.read_excel('file.xlsx')
- df = pd.read_excel('file.xlsx', sheet_name = 'first_sheet')

In [128]:
import os
os.getcwd()

'/Users/kelvin/Desktop/workspace/AI_Basic'

In [129]:
# 파일 로딩 (hk_221206.csv), 절대경로/상대경로 (.\\, ..\\)
# df_hk = pd.read_csv('.\\data\\hk_221206.csv') # df_hk 라는 dataFrame명을 생성
df_hk = pd.read_csv('./ktd_2_data230224/hk_221206.csv')

In [130]:
# df_hk
df_hk

name    jumin7 gender  height  age blood_type company grades  salary  \
0      hkd1  990623-2      F   161.9   22          A       A      A    4100   
1     hkd10  900303-2      F   169.4   31          A       A      B    4720   
2    hkd100  681205-2      F   168.3   55          A       A      B    7280   
3    hkd101  931226-2      F   155.3   28         AB       B      B    4060   
4    hkd102  920123-1      M   188.6   29          O       B      F    4390   
..      ...       ...    ...     ...  ...        ...     ...    ...     ...   
245   hkd95  730710-2      F   157.8   50          A       A      A    6990   
246   hkd96  720805-2      F   150.4   51          O       A      A    7110   
247   hkd97  710929-1      M   172.4   52          A       A      B    8000   
248   hkd98  701024-1      M   166.7   53          O       A      B    8760   
249   hkd99  691115-2      F   152.6   54          B       A      B    7470   

     expenditure  
0           1975  
1           2970  
2           5905  
3           2935  
4           4015  
..           ...  
245         4865  
246         4985  
247         6625  
248         7385  
249         6095  

[250 rows x 10 columns]

In [131]:
df_hk.head()

name    jumin7 gender  height  age blood_type company grades  salary  \
0    hkd1  990623-2      F   161.9   22          A       A      A    4100   
1   hkd10  900303-2      F   169.4   31          A       A      B    4720   
2  hkd100  681205-2      F   168.3   55          A       A      B    7280   
3  hkd101  931226-2      F   155.3   28         AB       B      B    4060   
4  hkd102  920123-1      M   188.6   29          O       B      F    4390   

   expenditure  
0         1975  
1         2970  
2         5905  
3         2935  
4         4015

**head()와 tail()**

head()는 맨 앞부터 데이터 추출 (기본5개),    
tail()은 맨 뒤부터 데이터 추출 (기본5개)  

In [132]:
df_hk.tail()

name    jumin7 gender  height  age blood_type company grades  salary  \
245  hkd95  730710-2      F   157.8   50          A       A      A    6990   
246  hkd96  720805-2      F   150.4   51          O       A      A    7110   
247  hkd97  710929-1      M   172.4   52          A       A      B    8000   
248  hkd98  701024-1      M   166.7   53          O       A      B    8760   
249  hkd99  691115-2      F   152.6   54          B       A      B    7470   

     expenditure  
245         4865  
246         4985  
247         6625  
248         7385  
249         6095

**DataFrame 출력 시 option**
* max_rows, max_columns, max_colwidth 설정


In [139]:
pd.set_option('display.max_rows',1000)
pd.set_option('display.max_columns',1000)
pd.set_option('display.max_colwidth',1000)
df_hk
pd.set_option('display.max_rows',50)
pd.set_option('display.max_columns',50)
pd.set_option('display.max_colwidth',50)
df_hk

name    jumin7 gender  height  age blood_type company grades  salary  \
0      hkd1  990623-2      F   161.9   22          A       A      A    4100   
1     hkd10  900303-2      F   169.4   31          A       A      B    4720   
2    hkd100  681205-2      F   168.3   55          A       A      B    7280   
3    hkd101  931226-2      F   155.3   28         AB       B      B    4060   
4    hkd102  920123-1      M   188.6   29          O       B      F    4390   
..      ...       ...    ...     ...  ...        ...     ...    ...     ...   
245   hkd95  730710-2      F   157.8   50          A       A      A    6990   
246   hkd96  720805-2      F   150.4   51          O       A      A    7110   
247   hkd97  710929-1      M   172.4   52          A       A      B    8000   
248   hkd98  701024-1      M   166.7   53          O       A      B    8760   
249   hkd99  691115-2      F   152.6   54          B       A      B    7470   

     expenditure  
0           1975  
1           2970  
2           5905  
3           2935  
4           4015  
..           ...  
245         4865  
246         4985  
247         6625  
248         7385  
249         6095  

[250 rows x 10 columns]

## Quiz 2

1) bike.csv를 읽고 column이 몇 개 인지 기술 하시오
12

2) iris.cvs를 읽고 row가 몇 개 인지 기술 하시오 
150

3) Online2.txt를 읽고 3번째 column이 무엇인지 기술 하시오 
Description

In [146]:
df_bike = pd.read_csv('./ktd_2_data230224/bike.csv')
df_iris = pd.read_csv('./ktd_2_data230224/iris.csv')
df_Online2 = pd.read_csv('./ktd_2_data230224/Online2.txt',sep='\t')
df_bike
df_iris
df_Online2

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

       InvoiceDate  UnitPrice  CustomerID         Country  
0       2010-12-01       2.55     17850.0  United Kingdom  
1       2010-12-01       3.39     17850.0  United Kingdom  
2       2010-12-01       2.75     17850.0  United Kingdom  
3       2010-12-01       3.39     17850.0  United Kingdom  
4       2010-12-01       3.39     17850.0  United Kingdom  
...            ...        ...         ...             ...  
541904  2011-12-09       0.85     12680.0          France  
541905  2011-12-09       2.10     12680.0          France  
541906  2011-12-09       4.15     12680.0          France  
541907  2011-12-09       4.15     12680.0          France  
541908  2011-12-09       4.95     12680.0          France  

[541909 rows x 8 columns]

### 2_3. DataFrame 속성(attribute)
* df.ndim: 객체의 차원
* df.shape: 각 차원의 길이
* df.dtypes: 데이터 타입 dtype와 구별, 복수형이라야 함
* df.info(): 데이터 타입 정보 요약, 결측치 정보
* df.columns: column명
* df.describe(): 수치형 데이터 요약 정보

In [147]:
# df.ndim: 객체의 차원
df_hk.ndim

2

In [148]:
# df.shape: 각 차원의 길이
df_hk.shape

(250, 10)

In [149]:
# df.dtypes: 데이터 타입 dtype와 구별, 복수형이라야 함
df_hk.dtypes

name            object
jumin7          object
gender          object
height         float64
age              int64
blood_type      object
company         object
grades          object
salary           int64
expenditure      int64
dtype: object

In [150]:
# df.info(): 데이터 타입 정보 요약, 결측치 정보
df_hk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         250 non-null    object 
 1   jumin7       250 non-null    object 
 2   gender       250 non-null    object 
 3   height       250 non-null    float64
 4   age          250 non-null    int64  
 5   blood_type   250 non-null    object 
 6   company      250 non-null    object 
 7   grades       250 non-null    object 
 8   salary       250 non-null    int64  
 9   expenditure  250 non-null    int64  
dtypes: float64(1), int64(3), object(6)
memory usage: 19.7+ KB


In [151]:
# df.columns: column명
df_hk.columns

Index(['name', 'jumin7', 'gender', 'height', 'age', 'blood_type', 'company',
       'grades', 'salary', 'expenditure'],
      dtype='object')

In [152]:
# df.describe(): 수치형 데이터 요약 정보
df_hk.describe()

height         age        salary   expenditure
count  250.000000  250.000000    250.000000    250.000000
mean   169.569200   39.240000   6357.480000   4971.480000
std     12.283857    8.653899   1849.791383   1861.112508
min    148.100000   20.000000   3080.000000   1330.000000
25%    159.300000   33.000000   5002.500000   3593.750000
50%    168.750000   39.000000   6100.000000   4762.500000
75%    180.375000   46.000000   7427.500000   6272.500000
max    191.800000   55.000000  11990.000000  10865.000000

In [154]:
# df.describe(): all 데이터 요약 정보
df_hk.describe()

height         age        salary   expenditure
count  250.000000  250.000000    250.000000    250.000000
mean   169.569200   39.240000   6357.480000   4971.480000
std     12.283857    8.653899   1849.791383   1861.112508
min    148.100000   20.000000   3080.000000   1330.000000
25%    159.300000   33.000000   5002.500000   3593.750000
50%    168.750000   39.000000   6100.000000   4762.500000
75%    180.375000   46.000000   7427.500000   6272.500000
max    191.800000   55.000000  11990.000000  10865.000000

### 2_4. Index와 Column 객체

In [156]:
# 파일 로딩 
df_hk = pd.read_csv('./ktd_2_data230224/hk_221206.csv')
df_hk.head(2)

name    jumin7 gender  height  age blood_type company grades  salary  \
0   hkd1  990623-2      F   161.9   22          A       A      A    4100   
1  hkd10  900303-2      F   169.4   31          A       A      B    4720   

   expenditure  
0         1975  
1         2970

In [157]:
# Index, column 보기
df_hk.index.values[:50]

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])

In [158]:
# Index 객체 추출
df_hk.index

RangeIndex(start=0, stop=250, step=1)

In [ ]:
# Index 객체 추출 (전체)
df_hk.index.values

In [ ]:
# Index 객체 추출 (일부)
df_hk.index.values[:50]

In [159]:
# column 객체 추출
df_hk.columns

Index(['name', 'jumin7', 'gender', 'height', 'age', 'blood_type', 'company',
       'grades', 'salary', 'expenditure'],
      dtype='object')

In [161]:
# column 객체 추출 (일부)
df_hk.columns.values[:50]

array(['name', 'jumin7', 'gender', 'height', 'age', 'blood_type',
       'company', 'grades', 'salary', 'expenditure'], dtype=object)

In [165]:
# column 내용 추출 (1개)
df_hk['age']

0      22
1      31
2      55
3      28
4      29
       ..
245    50
246    51
247    52
248    53
249    54
Name: age, Length: 250, dtype: int64

In [170]:
# column 내용 추출 (2개 이상)
df_hk[['age','jumin7']]

age    jumin7
0     22  990623-2
1     31  900303-2
2     55  681205-2
3     28  931226-2
4     29  920123-1
..   ...       ...
245   50  730710-2
246   51  720805-2
247   52  710929-1
248   53  701024-1
249   54  691115-2

[250 rows x 2 columns]

## Quiz 3

1) bike.csv의 column을 전부 기술 하시오

2) iris.cvs의 정수형 변수는 몇 개 인지 기술 하시오

In [171]:
df_bike.columns
df_iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


### 2_5. 함수 및 메소드
* df['age'].mean(): 평균값
* df['age'].median(): 중앙값
* df['age'].min(): 최소값
* df['age'].max(): 최대값
* df['age'].var(): 분산
* df['age'].std(): 표준편차
* df['age'].skew(): 왜도
* df['age'].kurt(): 첨도
* df['age'].count(): 개체수
* df['age'].quantile(q = 0.25) : 1사분위수

In [172]:
# 수치형변수 함수 및 메소드
print( "df_hk['age'].min() : ", df_hk['age'].min())
print( "df_hk['age'].max() : ", df_hk['age'].max())
print( "df_hk['age'].var() : ", df_hk['age'].var())
print( "df_hk['age'].std() : ", df_hk['age'].std())
print( "df_hk['age'].skew() : ", df_hk['age'].skew())
print( "df_hk['age'].kurt() : ", df_hk['age'].kurt())

df_hk['age'].min() :  20
df_hk['age'].max() :  55
df_hk['age'].var() :  74.88995983935743
df_hk['age'].std() :  8.653898534149649
df_hk['age'].skew() :  -0.07707060633056183
df_hk['age'].kurt() :  -0.7894949277134797


* df['blood_type'].nunique(): Column내 개별 데이터값 갯수
* df['blood_type'].unique(): Column내 개별 데이터값 내용
* df['blood_type'].value_counts(): Column내 개별 데이터값의 분포

  주의) default는 Null 값 제외, Null값 포함하려면 dropna를 True로 세팅

In [176]:
# 명목형변수 함수 및 메소드
df_hk['blood_type'].nunique()
df_hk['blood_type'].unique()
df_hk['blood_type'].value_counts()

A     84
O     83
B     54
AB    29
Name: blood_type, dtype: int64

## Quiz 4

1) iris.cvs의 column('species')의 value_counts()를 구하시오

2) Online2.txt의 column('Country')의 unique()를 구하시오

In [179]:
df_iris['species'].value_counts()
df_Online2['Country'].nunique()

38

### 2_6. column 생성, 수정, 삭제


In [180]:
# column 생성
df_hk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         250 non-null    object 
 1   jumin7       250 non-null    object 
 2   gender       250 non-null    object 
 3   height       250 non-null    float64
 4   age          250 non-null    int64  
 5   blood_type   250 non-null    object 
 6   company      250 non-null    object 
 7   grades       250 non-null    object 
 8   salary       250 non-null    int64  
 9   expenditure  250 non-null    int64  
dtypes: float64(1), int64(3), object(6)
memory usage: 19.7+ KB


In [182]:
# column 수정
df_hk['point'] = df_hk['age']+10
df_hk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         250 non-null    object 
 1   jumin7       250 non-null    object 
 2   gender       250 non-null    object 
 3   height       250 non-null    float64
 4   age          250 non-null    int64  
 5   blood_type   250 non-null    object 
 6   company      250 non-null    object 
 7   grades       250 non-null    object 
 8   salary       250 non-null    int64  
 9   expenditure  250 non-null    int64  
 10  point        250 non-null    int64  
dtypes: float64(1), int64(4), object(6)
memory usage: 21.6+ KB


**axis에 따른 삭제**

axis=1 일 경우 drop()은 column 방향으로 삭제

In [188]:
# 1개 column 삭제
df_hk.drop('point',axis=1, inplace=True)

KeyError: "['point'] not found in axis"

In [4]:
# 2개 이상 column 삭제는 list로 입력. inplace=True일 경우 호출을 한 DataFrame에 drop결과가 직접 반영 
import pandas as pd
df_hk = pd.read_csv('./ktd_2_data230224/hk_221206.csv')
df_drop=df_hk.drop(['age','height','salary'],axis=1)
df_drop

name    jumin7 gender blood_type company grades  expenditure
0      hkd1  990623-2      F          A       A      A         1975
1     hkd10  900303-2      F          A       A      B         2970
2    hkd100  681205-2      F          A       A      B         5905
3    hkd101  931226-2      F         AB       B      B         2935
4    hkd102  920123-1      M          O       B      F         4015
..      ...       ...    ...        ...     ...    ...          ...
245   hkd95  730710-2      F          A       A      A         4865
246   hkd96  720805-2      F          O       A      A         4985
247   hkd97  710929-1      M          A       A      B         6625
248   hkd98  701024-1      M          O       A      B         7385
249   hkd99  691115-2      F          B       A      B         6095

[250 rows x 7 columns]

axis=0 일 경우 drop()은 row 방향으로 삭제 

In [6]:
# 1개 row 삭제
df_drop.drop(0,axis=0)

name    jumin7 gender blood_type company grades  expenditure
1     hkd10  900303-2      F          A       A      B         2970
2    hkd100  681205-2      F          A       A      B         5905
3    hkd101  931226-2      F         AB       B      B         2935
4    hkd102  920123-1      M          O       B      F         4015
5    hkd103  910215-2      F          A       B      A         1915
..      ...       ...    ...        ...     ...    ...          ...
245   hkd95  730710-2      F          A       A      A         4865
246   hkd96  720805-2      F          O       A      A         4985
247   hkd97  710929-1      M          A       A      B         6625
248   hkd98  701024-1      M          O       A      B         7385
249   hkd99  691115-2      F          B       A      B         6095

[249 rows x 7 columns]

In [7]:
# 2개 이상 row 삭제
df_drop.drop([0,1],axis=0)

name    jumin7 gender blood_type company grades  expenditure
2    hkd100  681205-2      F          A       A      B         5905
3    hkd101  931226-2      F         AB       B      B         2935
4    hkd102  920123-1      M          O       B      F         4015
5    hkd103  910215-2      F          A       B      A         1915
6    hkd104  990623-1      M          A       B      A         2375
..      ...       ...    ...        ...     ...    ...          ...
245   hkd95  730710-2      F          A       A      A         4865
246   hkd96  720805-2      F          O       A      A         4985
247   hkd97  710929-1      M          A       A      B         6625
248   hkd98  701024-1      M          O       A      B         7385
249   hkd99  691115-2      F          B       A      B         6095

[248 rows x 7 columns]

column명 변경

In [9]:
# rename
df_hk.rename(columns={'gender':'sex'})

name    jumin7 sex  height  age blood_type company grades  salary  \
0      hkd1  990623-2   F   161.9   22          A       A      A    4100   
1     hkd10  900303-2   F   169.4   31          A       A      B    4720   
2    hkd100  681205-2   F   168.3   55          A       A      B    7280   
3    hkd101  931226-2   F   155.3   28         AB       B      B    4060   
4    hkd102  920123-1   M   188.6   29          O       B      F    4390   
..      ...       ...  ..     ...  ...        ...     ...    ...     ...   
245   hkd95  730710-2   F   157.8   50          A       A      A    6990   
246   hkd96  720805-2   F   150.4   51          O       A      A    7110   
247   hkd97  710929-1   M   172.4   52          A       A      B    8000   
248   hkd98  701024-1   M   166.7   53          O       A      B    8760   
249   hkd99  691115-2   F   152.6   54          B       A      B    7470   

     expenditure  
0           1975  
1           2970  
2           5905  
3           2935  
4           4015  
..           ...  
245         4865  
246         4985  
247         6625  
248         7385  
249         6095  

[250 rows x 10 columns]

In [ ]:
# df.columns=['name','jumin7']

## Quiz 5

1) iris.cvs의 column('sepal_length' , 'sepal_width')의 숫자 합을  파생변수 'sepal'로 생성하시오

In [10]:
df_iris = pd.read_csv('./ktd_2_data230224/iris.csv')
df_iris['sepal'] = df_iris['sepal_length'] + df_iris['sepal_width']

### 2_7. 인덱싱 및 필터링

* DataFrame [ ] 로 row와 column을 인덱싱

In [11]:
# 파일 로딩 
df_hk = pd.read_csv('./ktd_2_data230224/hk_221206.csv')

In [12]:
# []에 한개의 column만 입력하면 Series 객체를 반환  
df_hk[['age','jumin7']]

age    jumin7
0     22  990623-2
1     31  900303-2
2     55  681205-2
3     28  931226-2
4     29  920123-1
..   ...       ...
245   50  730710-2
246   51  720805-2
247   52  710929-1
248   53  701024-1
249   54  691115-2

[250 rows x 2 columns]

In [13]:
# []에 여러개 column을 list로 입력하면 column들로 구성된 DataFrame을 반환  
df_hk[2:5]

name    jumin7 gender  height  age blood_type company grades  salary  \
2  hkd100  681205-2      F   168.3   55          A       A      B    7280   
3  hkd101  931226-2      F   155.3   28         AB       B      B    4060   
4  hkd102  920123-1      M   188.6   29          O       B      F    4390   

   expenditure  
2         5905  
3         2935  
4         4015

In [14]:
# []에 한개 index 정보를 입력하면 DataFrame을 반환  
df_hk[:9]

name    jumin7 gender  height  age blood_type company grades  salary  \
0    hkd1  990623-2      F   161.9   22          A       A      A    4100   
1   hkd10  900303-2      F   169.4   31          A       A      B    4720   
2  hkd100  681205-2      F   168.3   55          A       A      B    7280   
3  hkd101  931226-2      F   155.3   28         AB       B      B    4060   
4  hkd102  920123-1      M   188.6   29          O       B      F    4390   
5  hkd103  910215-2      F   150.2   30          A       B      A    3290   
6  hkd104  990623-1      M   176.0   22          A       B      A    3750   
7  hkd105  980717-2      F   163.4   23          B       B      B    3080   
8  hkd106  970818-1      M   186.3   24         AB       B      C    3110   

   expenditure  
0         1975  
1         2970  
2         5905  
3         2935  
4         4015  
5         1915  
6         2375  
7         1330  
8         2360

In [ ]:
# []에 index 정보입력 오류


* iloc[ ] integer location : 위치기반 indexing

   [row정보, column정보] row, column 정보는 integer

In [21]:
# iloc
df_hk.iloc[0:3,0:5]

name    jumin7 gender  height  age
0    hkd1  990623-2      F   161.9   22
1   hkd10  900303-2      F   169.4   31
2  hkd100  681205-2      F   168.3   55

* loc[ ] location : 명칭기반 indexing (slice indexing)

   [row정보, column정보] column 정보는 column명

In [19]:
# loc
df_hk.loc[0:3, 'name':'age']

name    jumin7 gender  height  age
0    hkd1  990623-2      F   161.9   22
1   hkd10  900303-2      F   169.4   31
2  hkd100  681205-2      F   168.3   55
3  hkd101  931226-2      F   155.3   28

* boolean indexing

## Quiz 6

1) iris.cvs의 column('petal_width') 값이 1.0보다 큰 것은 몇개인지 구하시오

2) bike.cvs의 column dtype이 'int64' 인것만 추출하고 DataFrame명을 'df_bike_int64'로 저장 하시오 

In [23]:
df_iris.loc[df_iris['petal_width']>1.0, :]

sepal_length  sepal_width  petal_length  petal_width     species  sepal
50            7.0          3.2           4.7          1.4  versicolor   10.2
51            6.4          3.2           4.5          1.5  versicolor    9.6
52            6.9          3.1           4.9          1.5  versicolor   10.0
53            5.5          2.3           4.0          1.3  versicolor    7.8
54            6.5          2.8           4.6          1.5  versicolor    9.3
..            ...          ...           ...          ...         ...    ...
145           6.7          3.0           5.2          2.3   virginica    9.7
146           6.3          2.5           5.0          1.9   virginica    8.8
147           6.5          3.0           5.2          2.0   virginica    9.5
148           6.2          3.4           5.4          2.3   virginica    9.6
149           5.9          3.0           5.1          1.8   virginica    8.9

[93 rows x 6 columns]

In [33]:

df_bike_int64= df_bike.loc[:,df_bike.dtype == 'int64']

AttributeError: 'DataFrame' object has no attribute 'dtype'

## Quiz 7

(실습문제 유형1을 사용하여, : indian_food2.csv (구분자: 쉼표(,), 255 행, 10 컬럼)

1) 단계 1-1. 지역(state) 컬럼을 제거하시오.

2) 단계 1-2. 조리 시간(cook_time)과 재료 준비 시간(prep_time)이 양수(>0)인 데이터만 추출하시오.

### 2_8. 결손 데이터 처리하기
* isna(), dropna(), fillna()

In [ ]:
# 파일 로딩 (hk_221206_na.csv)


In [ ]:
# isna()


In [ ]:
# dropna()


In [ ]:
# fillna()


## Quiz 8

(실습문제 유형5을 사용하여, : phone.csv (구분자:  “,” , 105,828 Rows, 10 Columns)

1) 단계 1-1. 데이터 중 테스트 결과(TestResult)가 SUCCESS이고 신호 강도(Signal) 값이 결측치(null, N/A 등)가 아닌 데이터만 추출

### 2_9. sort, aggregation함수, groupby 

* sort_values()


In [ ]:
# 파일 로딩


In [ ]:
# sort_values, 1개 column 정렬

In [ ]:
# n개 column 정렬  ['blood_type', 'age']


In [ ]:
# n개 column 정렬, ascending(True/False) 적용


* aggregation (집계함수)

In [ ]:
# 1개 column 에 대한 aggregation


In [ ]:
# n개 column 에 대한 aggregation


In [ ]:
# n개 column 에 대한 n개 aggregation


In [ ]:
# n개 column 에 대한 column별 aggregation (dict형태로 입력)


In [ ]:
# 1개 column에 2개 이상 aggregation을 적용하면 마지막 aggregation이 적용


In [ ]:
# 1개 column에 2개 이상 aggregation 적용 방법


* groupby()

In [ ]:
# 1개 column 에 대한 groupby


In [ ]:
# n개 column 에 대한 groupby 


* groupby() + aggregation

In [ ]:
# groupby


In [ ]:
# n개 groupby + n개 aggregation(1개 column)


In [ ]:
# n개 groupby + n개 aggregation(column별 동일 aggregation)


In [ ]:
# n개 groupby + n개 aggregation(column별 다른 aggregation)


## Quiz 9

1) iris.cvs의 column('species')별로 'sepal_length' , 'sepal_width')의 평균, 중앙값, 최대값을 구하시오

### 2_10. DataFrame의 병합 (join)
* concat

In [ ]:
# 파일 로딩


In [ ]:
# concat 실습 파일 생성


In [ ]:
# concat (row), axis = 0 default 


In [ ]:
# concat (column)  axis = 1


* merge

In [ ]:
# merge 실습 파일 생성


In [ ]:
# merge  (inner join default)


In [ ]:
# merge   (left join)


In [ ]:
# merge   (outer join)


### 2_11. column내 값 변경
* replace
* map

In [ ]:
# 파일로딩

In [ ]:
# replace (1개 값 변경)


In [ ]:
# replace (n개 값 변경)


In [ ]:
# map (n개 값 변경)


### 2_12. apply lambda 식으로 데이터 가공

In [ ]:
# 파일 로딩 


In [ ]:
# apply lambda


### 2_13. 값 등급화, 분류
* np.where
* cut

In [ ]:
# np.where 2개
# np.where(조건, True, False)


In [ ]:
# np.where 3개
# np.where(조건1, True, False)
# np.where(조건1, True, 
#                 np.where(조건2, True, False)


In [ ]:
# cut
pd.cut(x = df_hk['age'], bins = [0,20,30,40,100], labels = ['10대', '20대', '30대', '40대이상'])

### 2_14. 날짜 변수 처리 (datetime)

In [ ]:
import datetime

In [ ]:
# pd.to_datetime


In [ ]:
# astype('datetime64')


In [ ]:
# datetime.datetime.strptime
df_hk['jumin7'].apply(lambda x : x[0:6]).apply(lambda x : datetime.datetime.strptime(x, '%y%m%d'))

In [ ]:
# datetime 활용


Quiz 7 (실습문제 유형5을 사용하여, : phone.csv (구분자: “,” , 105,828 Rows, 10 Columns)

1) 단계 1-1. 데이터 중 테스트 결과(TestResult)가 SUCCESS이고 신호 강도(Signal) 값이 결측치(null, N/A 등)가 아닌 데이터만 추출

In [ ]:
# 파일 로딩


In [ ]:
# 단계 1-1. 데이터 중 테스트 결과(TestResult)가 SUCCESS


In [ ]:
# 신호 강도(Signal) 값이 결측치(null, N/A 등)가 아닌 데이터만 추출
